omnilib/control.lua

```diff
@@ -1,408 +1,343 @@
+	end
+	-- Update according to compression research status
+	if not recipe_tree and 
+		recipe.force.technologies["compression-recipes"] and
+		recipe.force.technologies["compression-recipes"].researched and 
+		not recipe.category:find("-compressed", nil, true) and 
+		not name:find("concentrated", nil, true) then
+		local recipes = recipe.force.recipes
+		-- Handle both compressed and permuted recipes.
+		local compressed, permuted_compressed = recipes[name.."-compression"], recipes[name:gsub("omniperm","compression-omniperm")]
+		if compressed then
+			compressed.enabled = true
+		end
+		if permuted_compressed then
+			permuted_compressed.enabled = true
+		end
+		-- Deal with generator variants
+		for tier=1, math.huge do
+			local compressed_recipe = recipes[name.."-concentrated-grade-"..tier]
+			if compressed_recipe then
+				compressed_recipe.enabled = true
+			else
+				break
 			end
-			if found_it then return true end
-		else
-			if string.find(str,list[i]) then return true end
 		end
 	end
-	return false
+	if recipe.enabled then
+		update_last_tier(recipe)
+	end
 end
-local initilizing = false
-function omni_update(game)
-	initilizing = true
-	log("running through omnidate")
-  for _, force in pairs(game.forces) do
-    log("researching compressed techs where the standard is done.")
-    for _,tech in pairs(force.technologies) do
-			if not start_with(tech.name,"omnipressed-") and tech.force.technologies["omnipressed-" .. tech.name] then
-				tech.force.technologies["omnipressed-" .. tech.name].researched = tech.researched
-      elseif start_with(tech.name,"omnipressed-") and tech.force.technologies[string.sub(tech.name,13)] then
-        tech.force.technologies[string.sub(tech.name,13)].researched = tech.researched
-      end
-    end
-    
-		if force.technologies["compression-recipes"] and force.technologies["compression-recipes"].researched then
-			log("enabling compressed recipes of researched techs")
-      for _, rec in pairs(force.recipes) do
-				if not string.find(rec.name,"compression") and rec.enabled then
-					local compressCheck = rec.name.."-compression"
-					if string.find(rec.name,"omniperm") then
-						local b=string.find(rec.name,"omniperm")
-            compressCheck=string.sub(rec.name,1,b-1).."compression-"..string.sub(rec.name,b,-1)
-          end
-					if force.recipes[compressCheck] then
-						force.recipes[compressCheck].enabled = true
-						local z=1
-						while force.recipes[rec.name.."-compression-grade-"..z] do
-							force.recipes[rec.name.."-compression-grade-"..z].enabled=true
-							z=z+1
+
+local building_tiers = {
+	"compact",
+	"nanite",
+	"quantum",
+	"singularity"
+}
+
+local function update_tech(tech)
+	local variant = tech.force.technologies["omnipressed-" .. tech.name] or tech.force.technologies[tech.name:gsub("^omnipressed%-", "")] or {}
+	-- Handle compressed techs
+	if tech.researched or variant.researched then
+		tech.researched, variant.researched = true, true
```

Comment:
how is this not crashing on multi-level research, like infinite tech? the version that was in the compression control file specifically asked for level for a reason ;)

Reply:
7efa654 should address your concerns. It didn't crash but it wasn't properly syncing them.